### ISSUE TITLE
How to get the learning rate in optimizer during training?

### ISSUE THREAD

**Issue by zhangjh915** [tensorflow/tensorflow#39782]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): TF 2.0
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
When I am training the model with `tf.keras.optimizers.Adam` optimizer, I want to extract the learning rate during training to add it into the tensorboard. I did not find any API for this function. And the only result I found on Google is https://stackoverflow.com/questions/55622400/can-you-extract-the-current-learning-rate-from-tf-keras-adam, which suggested to calculate it by myself. I wonder if such API exists or will be added in the future.

**Will this change the current api? How?**
No. May add new features.

**Who will benefit with this feature?**
Potentially everyone.

**Any Other info.**

**Comment by ymodak** [comment]
You may try https://github.com/keras-team/keras/issues/2823#issuecomment-221895779 in this case.

**Comment by zhangjh915** [comment]
> You may try [keras-team/keras#2823 (comment)](https://github.com/keras-team/keras/issues/2823#issuecomment-221895779) in this case.

This is basically calculating the learning rate by ourselves and is duplicate with the code inside the optimizer. And it also relies on correct math when writing the code. Why not adding an API to extract the learning rate directly instead of writing extra code?

**Comment by luozhouyang** [comment]
Hi, @zhangjh915 
You misunderstand that case. It's a custom lr scheduler, so we need to calculate the new lr according to training steps or epochs.
If you just need log the lr to tensorboard, a `tf.keras.callbacks.Callback` can be useful.

```python
class LearningRateLoggingCallback(tf.keras.callbacks.Callback):

    def on_epoch_end(self, epoch):
        lr = self.model.optimizer.lr
        tf.summary.scalar('learning rate', data=lr, step=epoch)
```

Here has an example: [TensorBoard Scalars: Logging training metrics in Keras](https://www.tensorflow.org/tensorboard/scalars_and_keras)

**Comment by marko-radojcic** [comment]
It might be a meta algorithm. Feeding the feedback loop.

**Comment by zhangjh915** [comment]
> Hi, @zhangjh915
> You misunderstand that case. It's a custom lr scheduler, so we need to calculate the new lr according to training steps or epochs.
> If you just need log the lr to tensorboard, a `tf.keras.callbacks.Callback` can be useful.
> 
> ```python
> class LearningRateLoggingCallback(tf.keras.callbacks.Callback):
> 
>     def on_epoch_end(self, epoch):
>         lr = self.model.optimizer.lr
>         tf.summary.scalar('learning rate', data=lr, step=epoch)
> ```
> 
> Here has an example: [TensorBoard Scalars: Logging training metrics in Keras](https://www.tensorflow.org/tensorboard/scalars_and_keras)

I am using a customer training loop so I cannot use callback. And if I directly extract the learning rate from the optimizer, say optimizer.lr, it is always the initial learning rate which I define the optimizer instead of the varying learning rate during training.

**Comment by minhhoangbui** [comment]
@zhangjh915 This is what I need too, I htop there is an API which helps us to retrieve this param

**Comment by noirmist** [comment]
@zhangjh915 @minhhoangbui 
From this code, you can get the decayed learning rate during training.
```
print(optimizer._decayed_lr('float32').numpy())
```

**Comment by ammarchalifah** [comment]
Are there any update on this issue? I desperately need an API to log actual learning rate and writing the calculation having big risk of math error.

**Comment by zhangjh915** [comment]
> Are there any update on this issue? I desperately need an API to log actual learning rate and writing the calculation having big risk of math error.

Problem solved. Try what @noirmist said.

**Comment by marcojulioarg** [comment]
Hello

How can I log the learning rate if using tf.keras.optimizers.schedules.ExponentialDecay

**Comment by FruVirus** [comment]
@marcojulioarg 

in general, you should be able to do this:

```
if isinstance(optimizer.lr, tf.keras.optimizers.schedules.LearningRateSchedule):
    current_lr = optimizer.lr(optimizer.iterations.numpy())
else:
    current_lr = optimizer.lr
```

Learning rate schedulers have to define a `__call__` method that takes a `step` argument. You can get the updated training step using `optimizer.iterations`---this keeps track over epochs as well.
